### DISCUSSION TITLE
LOTW Certificate Import

### DISCUSSION THREAD

**Discussion by jeffreycentex** [magicbug/Cloudlog#1526] [General]
Maybe someone out there can give me a hint.  I'm running the master branch of ClouldLog on Ubuntu 22.04LTS with PHP 7.4.

Everything is fine except for import my P12 cert.  It gets uploaded to the ./uploads/lotw/cert folder, but I just get returned to screen where it says I need to import a certificate.

Haven't found anything here or in issues related to this.

**Comment by magicbug** [comment]
Hmm, is the p12 password protected if so the system will just dump it, if not then bit more debugging required I'd suggest making sure everything in Admin -> Debug is green, might need to turn on the logging options in Codeigniter to see whats going on deeper in the code

**Reply by jeffreycentex** [reply]
Yep.  Everything there is green.  I've looked through the code and don't see where the problem lies.  I loaded the certificate on my computer without a password to verify that no password is found.

I wonder if this might be openssl related.   I get an error when I run the update_clublog_scp URL:

A PHP Error was encountered
Severity: Warning

Message: file_put_contents(./updates/status.html): failed to open stream: Permission denied

Filename: controllers/Update.php

Line Number: 206

A PHP Error was encountered
Severity: Warning

Message: gzopen(): SSL operation failed with code 1. OpenSSL Error messages: error:0A000126:SSL routines::unexpected eof while reading

Filename: controllers/Update.php

Line Number: 238

A PHP Error was encountered
Severity: Warning

Message: file_put_contents(./updates/status.html): failed to open stream: Permission denied

Filename: controllers/Update.php

Line Number: 206

**Reply by phl0** [reply]
Looks more like a problem with permissions for the user the webserver runs under. Seems like it cannot create some files.

**Reply by cyfrax** [reply]
same thing here. P12 file gets uploaded to uploads/lotw/certs but nothing happens - it doesn't show in the lotw page
73
DL2CC

**Reply by phl0** [reply]
Any errors in the web server log?

**Comment by cyfrax** [comment]
No errors unfortunately.

**Reply by phl0** [reply]
Neither cloudlog logs nor webservers's logs?!

**Reply by cyfrax** [reply]
sorry, where do I find the cloudlog logs?

**Reply by cyfrax** [reply]
application/logs/? that directory doesn't contain log files so far

**Reply by magicbug** [reply]
Unless you turn on debugging in /application/config/config.php it wont

**Comment by cyfrax** [comment]
OK, set $config['log_threshold'] = 4;
in config.php but application/logs/ still empty after trying to upload a lotw cert.
anywhere else to enable debug logging?

**Comment by jeffreycentex** [comment]
Here is what I get there:

INFO - 2022-09-19 12:55:51 --> Helper loaded: form_helper
INFO - 2022-09-19 12:55:51 --> Language file loaded:
language/english/lotw_lang.php
INFO - 2022-09-19 12:55:51 --> Model Class Initialized
INFO - 2022-09-19 12:55:51 --> Model Class Initialized
INFO - 2022-09-19 12:55:51 --> Upload Class Initialized
INFO - 2022-09-19 12:55:51 --> Model Class Initialized
*ERROR - 2022-09-19 12:55:51 --> Severity: Notice --> Undefined index: cert
/var/www/html/application/controllers/Lotw.ph*
*p 392*
*ERROR - 2022-09-19 12:55:51 --> error:0308010C:digital envelope
routines::unsupported*
INFO - 2022-09-19 12:55:51 --> Config Class Initialized
INFO - 2022-09-19 12:55:51 --> Hooks Class Initialized
DEBUG - 2022-09-19 12:55:51 --> UTF-8 Support Enabled
INFO - 2022-09-19 12:55:51 --> Utf8 Class Initialized
INFO - 2022-09-19 12:55:51 --> URI Class Initialized
INFO - 2022-09-19 12:55:51 --> Router Class Initialized
INFO - 2022-09-19 12:55:51 --> Output Class Initialized

My cert is one that does not have a password assigned to it.  However, the
very first time I did this a long time ago, it did and failed.
I cleared all of the old certs out of the upload folder.

**Reply by phl0** [reply]
Seems that is an issue concerning the SSL cert stuff. U have OpenSSL installed and enabled for PHP? This is a sample output from phpinfo() on my system:

![Screenshot from 2022-09-19 15-51-35](https://user-images.githubusercontent.com/7112907/191033378-3f4ce373-36ab-4537-853e-6233bde46a44.png)

**Comment by jeffreycentex** [comment]
I have the same as you, except that I have two directives listed beneath
it, each with no value:

openssl.cafile
openssl.capath

**Reply by phl0** [reply]
Ok so this is not the issue then .... :(

**Comment by jeffreycentex** [comment]
Fixed it..  Ubuntu 22.04 issue.

diff -uN /etc/ssl/openssl.cnf\~original /etc/ssl/openssl.cnf for Nextcloud
with OpenSSL 3 · GitHub
<https://gist.github.com/rdh27785/97210d439a280063bd768006450c435d>

Add the legacy provider to openssl.cnf

**Comment by cyfrax** [comment]
Yeah, that did the trick for me as well. Cert uploads and activates fine now. Thanks Jeffrey!!
As for the non working logging, seems it was a permission issue with www-data. 
I think it's because of missing permissions as per https://github.com/magicbug/Cloudlog/wiki/Installation - the log files directory should probably be added to the list of directories.

**Comment by jeffreycentex** [comment]
Actually, it is...

I do:
openssl pkcs -clcerts -nokeys -in certfile.p12 -out cert.crt

and get:
error:  0308010C: digital envelope
routines:innder_evp_generic_fetch:unsupported:../crypto/evp/evp_fetch.c:349:Global
default library context, Algorithm (RC2-40-CBC : 0), Properties ()

Did ARRL update the certificate settings for new certificates?

**Comment by jeffreycentex** [comment]
Nope...
